src/charm.py

```diff
@@ -252,6 +333,312 @@ def _on_nginx_route_broken(self, _: Any) -> None:
+            )
+
+    # This method is too complex but hard to simplify.
+    def _certificate_revoked(self, revoke_list: List[Union[str, None]]) -> None:  # noqa: C901
+        """Handle TLS Certificate revocation.
+
+        Args:
+            revoke_list: TLS Certificates list to revoke
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
+        for hostname in revoke_list:
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}", tls_certificates_relation, self.app
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.remove_all_revisions()
+                except SecretNotFoundError:
+                    LOGGER.warning("Juju secret for %s already does not exist", hostname)
+                    continue
+            try:
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    tls_certificates_relation,
+                    self.app,
+                )
+                peer_relation = self.model.get_relation("nginx-peers")
+                if peer_relation:
+                    self._tls.pop_relation_data_fields(
+                        [f"key-{hostname}", f"password-{hostname}"], peer_relation, self.app
+                    )
+            except KeyError:
+                LOGGER.warning("Relation data for %s already does not exist", hostname)
+            self.certificates.request_certificate_revocation(
+                certificate_signing_request=old_csr.encode()
+            )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handle the TLS Certificate invalidation event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        if event.reason == "revoked":
+            hostname = self._tls.get_hostname_from_csr(
+                tls_certificates_relation, self.app, event.certificate_signing_request
+            )
+            self._certificate_revoked([hostname])
+        if event.reason == "expired":
+            self._on_certificate_expiring(event)
```

Comment:
I'd rather avoid calling a hoot method directly. Put the common business logic of both event in a private method and call it from both hook method, like you did with `_certificate_revoked`